src/crimson/thread/ThreadPool.h

```diff
@@ -0,0 +1,114 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*- 
+// vim: ts=8 sw=2 smarttab
+#pragma once
+
+#include <atomic>
+#include <condition_variable>
+#include <tuple>
+#include <type_traits>
+#include <boost/lockfree/queue.hpp>
+#include <boost/optional.hpp>
+#include <core/future.hh>
+#include <core/gate.hh>
+#include <core/semaphore.hh>
+#include <core/sharded.hh>
+
+#include "Condition.h"
+
+namespace ceph::thread {
+
+struct WorkItem {
+  virtual ~WorkItem() {}
+  virtual void process() = 0;
+};
+
+template<typename Func, typename T = std::invoke_result_t<Func>>
+struct Task final : WorkItem {
+  Func func;
+  seastar::future_state<T> state;
+  ceph::thread::Condition on_done;
+public:
+  explicit Task(Func&& f)
+    : func(std::move(f))
+  {}
+  void process() override {
+    try {
+      state.set(func());
+    } catch (...) {
+      state.set_exception(std::current_exception());
+    }
+    on_done.notify();
+  }
+  seastar::future<T> get_future() {
+    return on_done.wait().then([this] {
+      return seastar::make_ready_future<T>(state.get0(state.get()));
```

Comment:
i think this will copy the value. can this take advantage of the `future_state::get() &&` overload? ie `state.get0(std::move(state).get())`